### PR TITLE
genesis: have the output of genesis be a checkpoint

### PR DESCRIPTION
**Pull request by bmwill**
This PR reworks the genesis build process in order to have the output of genesis be a Checkpoint, signed by all validators in the initial validator set. As a first step this PR does the following:
- Defines the Genesis Transaction 
- Implements the logic for how the Genesis Transaction will be executed by the execution engine
- Reworks the genesis creation process to have the output be a checkpoint

While the output of genesis after this PR will be a checkpoint, it is still not used or inserted into a node's storage. The output objects have their previous transaction field rewritten to be all zeros as is done today. In a follow up PR we'll take the final step and actually use this checkpoint and transaction instead of just inserting genesis objects into the database on startup.

### CODE REVIEW

File: crates/sui-adapter/src/execution_engine.rs
Comment (sblackshear) [inline crates/sui-adapter/src/execution_engine.rs:340]: This seems like the right thing to do here, but just noting that this will lead to some slightly odd behavior the first time a genesis object is used. E.g., if a we have a genesis coin that is `N` bytes, we'll end up paying `N` * `storage_cost` the first time we send a tx that uses the coin, and the `storage_rebate` for the coin will be `N * storage_cost` afterward.

Comment (lxfind) [reply crates/sui-adapter/src/execution_engine.rs:340]: I wonder if we should actually do the accounting properly by initializing the storage fund pool to be the total storage cost of all the genesis objects, and then put the correct storage_rebate in the genesis objects?

Comment (bmwill) [reply crates/sui-adapter/src/execution_engine.rs:340]: So this is a problem today with the current way genesis creates these objects, it just becomes a bit more apparent now. I think that we're still a ways away from genesis being in a state we want it in and I do think that this is something we could consider fixing.

File: crates/sui-types/src/messages.rs
Comment (lxfind) [inline crates/sui-types/src/messages.rs:173]: What's the reason to define a new GenesisObject instead of just using Object?

File: crates/sui-types/src/messages.rs
Comment (lxfind) [inline crates/sui-types/src/messages.rs:1000]: The creation of the advance epoch system transaction probably looks the same, could we merge the logic?

Comment (bmwill) [reply crates/sui-types/src/messages.rs:1000]: Yeah happy to do so

File: crates/sui-config/src/builder.rs
Comment (aschran) [inline crates/sui-config/src/builder.rs:102]: Rather than `Vec` in particular, could this be `IntoIterator` to match the signature of `extend`?

File: crates/sui-config/src/genesis.rs
Comment (aschran) [inline crates/sui-config/src/genesis.rs:473]: was this supposed to be temporary or did you mean to merge? (the formatting makes it look unintentional)

**Review by aschran — APPROVED** [review]
prob wait for at least one more approval before merging :)

File: crates/sui-config/src/genesis.rs
Comment (bmwill) [reply crates/sui-config/src/genesis.rs]: Yes this is intentional and will be cleaned up in a follow up PR.

**Pull request merged**
